das/gcs_storage_service.go

```diff
@@ -0,0 +1,137 @@
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+
+	googleStorage "cloud.google.com/go/storage"
+	flag "github.com/spf13/pflag"
+)
+
+type GCSStorageServiceConfig struct {
+	Enable                 bool   `koanf:"enable"`
+	Bucket                 string `koanf:"bucket"`
+	ObjectPrefix           string `koanf:"object-prefix"`
+	DiscardAfterTimeout    bool   `koanf:"discard-after-timeout"`
+	SyncFromStorageService bool   `koanf:"sync-from-storage-service"`
+	SyncToStorageService   bool   `koanf:"sync-to-storage-service"`
+}
+
+var DefaultGCSStorageServiceConfig = GCSStorageServiceConfig{}
+
+func GCSConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Bool(prefix+".enable", DefaultGCSStorageServiceConfig.Enable, "enable storage/retrieval of sequencer batch data from an Google cloud storage bucket")
+	f.String(prefix+".bucket", DefaultGCSStorageServiceConfig.Bucket, "Google cloud storage bucket")
+	f.String(prefix+".object-prefix", DefaultGCSStorageServiceConfig.ObjectPrefix, "prefix to add to GCS objects")
+	f.Bool(prefix+".discard-after-timeout", DefaultGCSStorageServiceConfig.DiscardAfterTimeout, "discard data after its expiry timeout")
+	f.Bool(prefix+".sync-from-storage-service", DefaultGCSStorageServiceConfig.SyncFromStorageService, "enable gcs to be used as a source for regular sync storage")
+	f.Bool(prefix+".sync-to-storage-service", DefaultGCSStorageServiceConfig.SyncToStorageService, "enable gcs to be used as a sink for regular sync storage")
+}
+
+type GCSStorageService struct {
+	client              *googleStorage.Client
+	bucket              string
+	objectPrefix        string
+	discardAfterTimeout bool
+}
+
+func NewGCSStorageService(config GCSStorageServiceConfig) (StorageService, error) {
+	// https://cloud.google.com/docs/authentication/provide-credentials-adc
+	client, err := googleStorage.NewClient(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+	return &GCSStorageService{
+		client:              client,
+		bucket:              config.Bucket,
+		objectPrefix:        config.ObjectPrefix,
+		discardAfterTimeout: config.DiscardAfterTimeout,
+	}, nil
+}
+
+func (gcs *GCSStorageService) GetByHash(ctx context.Context, key common.Hash) ([]byte, error) {
+	log.Trace("das.GCSStorageService.GetByHash", "key", pretty.PrettyHash(key), "this", gcs)
+
+	object := gcs.client.Bucket(gcs.bucket).Object(gcs.objectPrefix + EncodeStorageServiceKey(key))
+	rc, err := object.NewReader(ctx)
+
+	rv := make([]byte, 0)
+
+	if _, err := rc.Read(rv); err != nil {
+		return nil, err
+	}
+
+	return rv, err
+}
+
+func (gcs *GCSStorageService) Put(ctx context.Context, value []byte, timeout uint64) error {
+	logPut("das.GCSStorageService.Store", value, timeout, gcs)
+
+	object := gcs.client.Bucket(gcs.bucket).Object(gcs.objectPrefix + EncodeStorageServiceKey(dastree.Hash(value)))
+	object = object.If(googleStorage.Conditions{DoesNotExist: true})
+
+	if !gcs.discardAfterTimeout {
```

Comment:
The `discardAfterTimeout` feature is actually for deleting the data from the storage backend after the data expiry timeout is reached. I'm not very familiar with GCS, does it have a way to set objects to auto-delete after some time? S3 objects have this feature. I think maybe it can be done with Object Lifecycle Management: https://cloud.google.com/storage/docs/lifecycle

The idea behind the `discardAfterTimeout` feature is that committee members promise to hold the data only for the expiry period (promising to hold data forever is a infinitely expensive promise to make, that's why there's a limit). For now, however, until we implement snap-sync, the data still needs to be stored somewhere so that nodes can sync, either in mirrors or the committee members ignoring the timeout.

From https://docs.arbitrum.io/inside-anytrust#sequencer-committee-interaction
_When the Arbitrum sequencer produces a data batch that it wants to post using the Committee, it sends the batch's data, **along with an expiration time (normally three weeks in the future)** via RPC to all Committee members in parallel. Each Committee member stores the data in its backing store, indexed by the data's hash. Then the member signs the (hash, expiration time) pair using its BLS key, and returns the signature with a success indicator to the sequencer._

Reply:
GCP storage doesn't seem to be able to set lifecycle for a single object.